File: Ruby/2024/04.rb
input = File.read(ARGV[0]).lines.map &:strip

NEIGH = [
  [-1, -1],
  [-1, 0],
  [-1, 1],
  [0, -1],
  [0, 1],
  [1, -1],
  [1, 0],
  [1, 1],
]

w = input[0].length
h = input.length
puts (0...h).flat_map { |i|
  (0...w).map { |j|
    NEIGH
      .filter { |n| (0...h).include?(i + 3 * n[0]) && (0...w).include?(j + 3 * n[1]) }
      .count { |n| (0...4).map { input[i + _1 * n[0]][j + _1 * n[1]] }.join == "XMAS" }
  }
}.sum

puts (1...h-1).map { |i|
  (1...w-1).count { |j|
    input[i][j] == "A" \
      && Set[input[i-1][j-1],input[i+1][j+1]] == Set[?M,?S] \
      && Set[input[i-1][j+1],input[i+1][j-1]] == Set[?M,?S]
  }
}.sum

File: Rust/2024/04.rs
#![feature(test)]

type Input = Vec<Vec<u8>>;

const NEIGH: [(isize, isize); 8] = [
    (-1, -1),
    (-1, 0),
    (-1, 1),
    (0, -1),
    (0, 1),
    (1, -1),
    (1, 0),
    (1, 1),
];

fn setup(input: &str) -> Input {
    input.trim().lines().map(|l| l.bytes().collect()).collect()
}

fn part1(input: &Input) -> usize {
    let height = input.len() as isize;
    let width = input[0].len() as isize;
    (0..height)
        .flat_map(|i| {
            (0..width).flat_map(move |j| {
                NEIGH.iter().filter(move |&(di, dj)| {
                    "XMAS".bytes().enumerate().all(|(k, b)| {
                        let k = k as isize;
                        let i = i + di * k;
                        let j = j + dj * k;
                        (0..height).contains(&i)
                            && (0..width).contains(&j)
                            && input[i as usize][j as usize] == b
                    })
                })
            })
        })
        .count()
}

fn part2(input: &Input) -> usize {
    let height = input.len();
    let width = input[0].len();
    (1..height - 1)
        .flat_map(|i| {
            (1..width - 1).filter(move |&j| {
                const OUTER: [(u8, u8); 2] = [(b'M', b'S'), (b'S', b'M')];
                input[i][j] == b'A'
                    && OUTER.contains(&(input[i - 1][j - 1], input[i + 1][j + 1]))
                    && OUTER.contains(&(input[i - 1][j + 1], input[i + 1][j - 1]))
            })
        })
        .count()
}

aoc::main!(2024, 4, ex: 1);
